Configurations/shared-info.pl

```diff
@@ -0,0 +1,108 @@
+#! /usr/bin/env perl
+# -*- mode: perl; -*-
+# Copyright 2016-2017 The OpenSSL Project Authors. All Rights Reserved.
+#
+# Licensed under the OpenSSL license (the "License").  You may not use
+# this file except in compliance with the License.  You can obtain a copy
+# in the file LICENSE in the source distribution or at
+# https://www.openssl.org/source/license.html
+
+# This is a collection of extra attributes to be used as input for creating
+# shared libraries, currently on any Unix variant, including Unix like
+# environments on Windows.
+
+sub detect_gnu_ld {
+    my @lines =
+        `$config{cross_compile_prefix}$target{cc} -Wl,-V /dev/null 2>&1`;
+    return grep /^GNU ld/, @lines;
+}
+sub detect_gnu_cc {
+    my @lines =
+        `$config{cross_compile_prefix}$target{cc} -v 2>&1`;
+    return grep /gcc/, @lines;
```

Comment:
As pointed out in #4755, this is not reliable way to detect gcc.

Reply:
Ah...  ok, just gotta ensure %predefined is filled in early enough, then.

Reply:
... which isn't so easy, we have a bit of a catch 22 there, since `$target{cc}` is modified far later than these functions are called.  Note that his is copied directly from `Makefile.shared`, and also, this is only called for `svr5-shared`, where I'm unsure there's been any report of an issue (not that we get much reports from that platform anyway).  So I'd like to argue that we can leave this alone for this PR and solve in a more general in a separate effort (I know you're thinking of options).  Sounds reasonable to you?

Reply:
Here is a wild idea... Why can't those flags that are defined here in shared-info.pl be just a template? I mean why can't a target be described as `inherit_from => [ ..., "something-shared", ... ]`? Why is there additional level of "indirection" with shared_target "pointing" to separate data structure? Wouldn't it be sufficient if shared_target was rather just a "reference" to a *subroutine* that accepts complete target description as argument? And not dedicated data structure and target that is?

Reply:
'cause inherit_from is processed before the final target structure is done, and because the subs that are values for attributes only receive the inherited value for that same attribute. In the other PR, where I implement later processing of those subs, it may be possible to get any target value from any sub, that will make things easier...

Reply:
Forgot to say that some of those shared-info.pl structures depend on $target{cc}, so you can see how that's a problem to handle with inherit_from

Reply:
And the question is does it *have to* be that way? One can also put it in following way. This is kind of dominated by idea that it *has to* mimic Makefile.shared, but does it actually have to?